src/logger.ts

```diff
@@ -0,0 +1,432 @@
+import * as Bluebird from 'bluebird';
+import * as es from 'event-stream';
+import { ClientRequest } from 'http';
+import * as https from 'https';
+import * as _ from 'lodash';
+import * as Lock from 'rwlock';
+import * as stream from 'stream';
+import * as url from 'url';
+import * as zlib from 'zlib';
+
+import { EventTracker } from './event-tracker';
+import Docker = require('./lib/docker-utils');
+import { LogType } from './lib/log-types';
+
+const ZLIB_TIMEOUT = 100;
+const COOLDOWN_PERIOD = 5 * 1000;
+const KEEPALIVE_TIMEOUT = 60 * 1000;
+const RESPONSE_GRACE_PERIOD = 5 * 1000;
+
+const MAX_LOG_LENGTH = 10 * 1000;
+const MAX_PENDING_BYTES = 256 * 1024;
+
+interface Options extends url.UrlWithParsedQuery {
+	method: string;
+	headers: Dictionary<string>;
+}
+
+type LogMessage = Dictionary<any>;
+
+abstract class LogBackend {
+	public offlineMode: boolean;
+	public publishEnabled: boolean = true;
+
+	public abstract log(message: Dictionary<any>): void;
+}
+
+class ResinLogBackend extends LogBackend {
+
+	private req: ClientRequest | null = null;
+	private dropCount: number = 0;
+	private writable: boolean = true;
+	private gzip: zlib.Gzip | null = null;
+	private opts: Options;
+	private stream: stream.PassThrough;
+	timeout: NodeJS.Timer;
+
+	public constructor(
+		apiEndpoint: string,
+		uuid: string,
+		deviceApiKey: string,
+	) {
+		super();
+
+		this.opts = url.parse(`${apiEndpoint}/device/v2/${uuid}/log-stream`) as any;
+		this.opts.method = 'POST';
+		this.opts.headers = {
+			Authorization: `Bearer ${deviceApiKey}`,
+			'Content-Type': 'application/x-ndjson',
+			'Content-Encoding': 'gzip',
+		};
+
+		// This stream serves serves as a message buffer during reconnections
+		// while we unpipe the old, malfunctioning connection and then repipe a
+		// new one.
+		this.stream = new stream.PassThrough({
+			allowHalfOpen: true,
+
+			// We halve the high watermark because a passthrough stream has two
+			// buffers, one for the writable and one for the readable side. The
+			// write() call only returns false when both buffers are full.
+			highWaterMark: MAX_PENDING_BYTES / 2,
+		});
+
+		this.stream.on('drain', () => {
+			this.writable = true;
+			this.flush();
+			if (this.dropCount > 0) {
+				this.write({
+					message: `Warning: Suppressed ${this.dropCount} message(s) due to high load`,
+					timestamp: Date.now(),
+					isSystem: true,
+					isStdErr: true,
+				});
+				this.dropCount = 0;
+			}
+		});
+	}
+
+	public log(message: LogMessage) {
+		if (this.offlineMode || !this.publishEnabled) {
+			return;
+		}
+
+		if (!_.isObject(message)) {
+			return;
+		}
+
+		message = _.assign({
+			timestamp: Date.now(),
+			message: '',
+		}, message);
+
+		if (!(message.isSystem || message.serviceId != null)) {
```

Comment:
Also do we want to be silently dropping logs? (Both here and for the !isObject above)

Reply:
I was just trying to keep it 1:1 with the coffeescript, we can handle that in a separate PR if we want it: https://github.com/resin-io/resin-supervisor/pull/733/files#diff-0f68e03b8d0f7c87ec71275f52139ee9L137

Reply:
Ok